techmatters/hrm
# chore: Add contact filters to generateElasticsearchQuery for hrm-search-config package [CHI-2704]

## Description
This PR
- Factors out `ConditionWhereClauses` type into `hrm-domain/hrm-core/permissions/queryGenerators/types.ts`. This type is defined so we have a "common ground" to the shape the "clauses generators" must comply with. This is so we don't forget to add more clause-generating functions if we ever add more permission rules.
- Moves `sqlGenerators.ts` file into `hrm-domain/hrm-core/permissions/queryGenerators` folder to have all the "clause generating" code grouped.
- Implements `elasticsearchGenerators.ts` under `hrm-domain/hrm-core/permissions/queryGenerators`. This file defines how the clauses are generated and processed.
  - A single generated clause has the type `QueryDslQueryContainer`, which is basically an ES query that can be nested.
  - The generated set of clauses for a given permission config is of type `QueryDslQueryContainer[][]` (an array of arrays of ES queries). 
  - The above is a 1-1 mapping between a "permissions ConditionsSets" to ES queries:
    Given permission rule for `action` that looks like `[ [cond1, cond2], [cond3] ]`, then the generator will output something like `[ [generateESQuery(cond1), generateESQuery(cond2) ], [ generateESQuery(cond3) ] ]`. The semantics of this "set of sets" is that, the final query "search `term`" should match the documents in elasticsearch
    - If `term` matches, filtering documents by `generateESQuery(cond1)` AND `generateESQuery(cond2)` or
    - If `term` matches, filtering documents by `generateESQuery(cond3)`
  - The special case where the user does not have permissions at all is captured in the `FILTER_ALL_CLAUSE`
    `[ [ { bool: { must_not: { match_all: { } } } } ] ]`. When this clause is added as a `filter` in the ES query, it will leave out all the documents.
- Adds functionality to generate ES queries in `hrm-domain/packages/hrm-search-config/generateElasticsearchQuery.ts`
  - `generateESFilter` is a helper function that will generate ES query filters from more human-readable parameters. E.g. `generateESFilter({ type: 'term', term: 'something', field: 'field1'  })` will generate a `term` query that matches documents that contain `something` term in the field `field1`. The idea of this one is to avoid slipping ES specific syntax in the code base.
  - `generateTranscriptQueriesFromFilters` will generate "a set of sets of queries" based on the `searchTerm` and "permissions ConditionsSets" given (intended for `viewTranscript` action).
  - `generateContactsQueriesFromFilters` does the same as above, but for contacts. In this case, a query on contacts is the same as searching `searchTerm` on contact content plus the cartesian product between the permissions ConditionsSets from `viewContact` and `viewTranscript` actions matching `searchTerm` on contact transcript:
    Given permission rules for `viewContact` like `[ [c1], [c2] ]` and for `viewTranscript` like `[ [t1], [t2] ]`, the query on `searchTerm` will match contacts if:
    - `c1` is true AND
      - `searchTerm` matches the contact content, OR
      - `searchTerm` matches the contact transcript AND `t1` is true, OR
      - `searchTerm` matches the contact transcript AND `t2` is true
    - `c2` is true AND
      - `searchTerm` matches the contact content, OR
      - `searchTerm` matches the contact transcript AND `t1` is true, OR
      - `searchTerm` matches the contact transcript AND `t2` is true
- Adds code to handle filters at contactService layer, in `hrm-domain/hrm-core/contact/contactSearchIndex.ts`:
  - `buildSearchFilters` and `generateContactSearchFilters` are two functions used to define "how the search filters that the user provide" should be translated to ES queries.
  - `conditionWhereClauses` defines "how each possible permission rule applied to contacts" should be translated to ES queries, `listContactsPermissionClause` and `generateContactPermissionsFilters` generate the ES queries for permission rules (intended for `viewContact` and `viewTranscript` actions).

### Checklist
- [x] [Corresponding issue has been opened](https://tech-matters.atlassian.net/browse/CHI-2704)

### Verification steps
- Fill in the details of the elastic search node configuration (see parameter `/development/us-east-1/hrm/shared/elasticsearch_config` in AWS parameter store). Paste the details in `.env`, with name `ELASTICSEARCH_CONFIG='{ ... the node details ... }'`
- Build and start `npm run build && npm run start:light`.
- Hot the new endpoint, replacing your Flex token and searchTerm in the following query
  `curl -X POST 'https://hrm-development.tl.techmatters.org/v0/accounts/ACd8a2e89748318adf6ddff7df6948deaf/contacts/searchV2?limit=20&offset=0' -H "Content-Type: application/json" -d '{ "searchParameters": {"searchTerm": "asdasd"  } }' -H "Authorization: Bearer <your token>"`
  - You can see the generated query logged in the server running console.
  - Contrast the results (count and list of contacts) with ElasticSearch Cloud Console.
- Play around with search parameters, by adding `counselor`/`dateFrom`/`dateTo` to the `searchParameters` object in the request.
- Play around changing rules for `viewContact` and `viewTranscript` actions in the `dev.json` permission file (you need to recompile and re-run the service for this - `npm run build && npm run start:service`).

Some example queries generated from different configs:


### AFTER YOU MERGE

1. Cut a release tag using the GitHub workflow. Wait for it to complete and notify in the #aselo-deploys Slack channel.
2. Comment on the ticket with the release tag version AND any additional instructions required to configure an environment to test the changes.
3. Only then move the ticket into the QA column in JIRA

You are responsible for ensuring the above steps are completed. If you move a ticket into QA without advising what version to test, the QA team will assume the latest tag has the changes. If it does not, the following confusion is on you! :-P
